File: runtime/ios/Runtime/Net/AISSEClient.swift
// AISSEClient — Server-Sent Events consumer for POST /ai/turn.
//
// The relay emits one event per frame (`event: token|tool_call|
// tool_result|finished|error`) with a single `data:` JSON line.  We
// parse the byte stream into discrete `AIFrame`s and yield them as an
// `AsyncThrowingStream` so callers can `for try await frame in ...`.
//
// No third-party SSE lib — URLSession's `bytes(for:)` gives us a line
// stream that's good enough for the tiny grammar we emit.
import Foundation

enum AIFrame {
    case token(turnID: String, text: String)
    case toolCall(turnID: String, callID: String, name: String, argsJSON: String)
    case toolResult(turnID: String, callID: String, resultJSON: String, isError: Bool)
    case finished(turnID: String, stopReason: String, tokensIn: Int, tokensOut: Int)
    case error(turnID: String, message: String)
}

struct AITurnRequest: Encodable {
    let boxID: String
    let text: String

    enum CodingKeys: String, CodingKey {
        case boxID = "box_id"
        case text
    }
}

actor AISSEClient {
    private let baseURL: URL
    private let session: URLSession
    private let tokenProvider: @Sendable () -> String?

    init(
        baseURL: URL = NetConfig.relayBaseURL,
        session: URLSession = .shared,
        tokenProvider: @escaping @Sendable () -> String?
    ) {
        self.baseURL = baseURL
        self.session = session
        self.tokenProvider = tokenProvider
    }

    /// Streams frames for a single AI turn.  The stream finishes after a
    /// `.finished` or `.error` frame, or when the underlying connection
    /// closes.
    func turn(boxID: String, text: String) -> AsyncThrowingStream<AIFrame, Error> {
        AsyncThrowingStream { continuation in
            let task = Task {
                do {
                    var req = URLRequest(url: baseURL.appendingPathComponent("/ai/turn"))
                    req.httpMethod = "POST"
                    req.setValue("application/json", forHTTPHeaderField: "Content-Type")
                    req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                    if let token = tokenProvider() {
                        req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
                    }
                    req.httpBody = try JSONEncoder().encode(
                        AITurnRequest(boxID: boxID, text: text)
                    )

                    let (bytes, response) = try await session.bytes(for: req)
                    if let http = response as? HTTPURLResponse,
                       !(200..<300).contains(http.statusCode) {
                        // Drain a small chunk of body for the error message.
                        var buf = Data()
                        for try await b in bytes.prefix(2048) { buf.append(b) }
                        let msg = String(data: buf, encoding: .utf8) ?? "(no body)"
                        throw HTTPError.status(http.statusCode, msg)
                    }

                    var event: String = "message"
                    for try await line in bytes.lines {
                        if Task.isCancelled { break }
                        if line.isEmpty {
                            event = "message"
                            continue
                        }
                        if line.hasPrefix("event: ") {
                            event = String(line.dropFirst(7))
                            continue
                        }
                        if line.hasPrefix("data: ") {
                            let json = String(line.dropFirst(6))
                            if let frame = parse(event: event, json: json) {
                                continuation.yield(frame)
                                if case .finished = frame { break }
                                if case .error = frame { break }
                            }
                        }
                    }
                    continuation.finish()
                } catch {
                    continuation.finish(throwing: error)
                }
            }
            continuation.onTermination = { _ in task.cancel() }
        }
    }

    // MARK: - parse helpers

    private func parse(event: String, json: String) -> AIFrame? {
        guard let data = json.data(using: .utf8) else { return nil }
        let dict = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        let turnID = dict["turn_id"] as? String ?? ""
        switch event {
        case "token":
            return .token(turnID: turnID, text: dict["text"] as? String ?? "")
        case "tool_call":
            return .toolCall(
                turnID: turnID,
                callID: dict["call_id"] as? String ?? "",
                name: dict["name"] as? String ?? "",
                argsJSON: dict["args_json"] as? String ?? ""
            )
        case "tool_result":
            return .toolResult(
                turnID: turnID,
                callID: dict["call_id"] as? String ?? "",
                resultJSON: dict["result_json"] as? String ?? "",
                isError: dict["is_error"] as? Bool ?? false
            )
        case "finished":
            return .finished(
                turnID: turnID,
                stopReason: dict["stop_reason"] as? String ?? "",
                tokensIn: (dict["tokens_in"] as? Int) ?? 0,
                tokensOut: (dict["tokens_out"] as? Int) ?? 0
            )
        case "error":
            return .error(
                turnID: turnID,
                message: dict["error"] as? String ?? "unknown error"
            )
        default:
            return nil
        }
    }
}

File: appunvs/android/app/src/main/java/com/appunvs/runtime/net/AISSEClient.kt
// AISSEClient — Server-Sent Events consumer for POST /ai/turn.
//
// We read the response body line-by-line through OkHttp's `BufferedSource`
// rather than pulling in a dedicated SSE library — the relay's grammar
// is small (one `event:` + one `data:` per record, blank line separator)
// and we want the streamed-frame surface to look the same on both
// platforms.
//
// The flow is cold: collecting starts the HTTP request; the upstream
// connection is closed on cancellation or after the terminal frame.
package com.appunvs.runtime.net

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody

sealed class AIFrame {
    data class Token(val turnID: String, val text: String) : AIFrame()
    data class ToolCall(val turnID: String, val callID: String, val name: String, val argsJSON: String) : AIFrame()
    data class ToolResult(val turnID: String, val callID: String, val resultJSON: String, val isError: Boolean) : AIFrame()
    data class Finished(val turnID: String, val stopReason: String, val tokensIn: Int, val tokensOut: Int) : AIFrame()
    data class Err(val turnID: String, val message: String) : AIFrame()
}

class AISSEClient(
    private val http: OkHttpClient,
    private val baseURL: String = NetConfig.relayBaseURL,
    private val json: Json = RelayClient.json,
) {

    fun turn(boxID: String, text: String): Flow<AIFrame> = flow {
        val body = json.encodeToString(
            AITurnRequest.serializer(),
            AITurnRequest(boxID = boxID, text = text),
        ).toRequestBody("application/json".toMediaType())

        val req = Request.Builder()
            .url("$baseURL/ai/turn")
            .header("Accept", "text/event-stream")
            .post(body)
            .build()

        http.newCall(req).execute().use { resp ->
            if (!resp.isSuccessful) {
                val msg = resp.body?.string()?.take(2048) ?: "(no body)"
                throw IllegalStateException("HTTP ${resp.code}: $msg")
            }
            val source = resp.body?.source() ?: throw IllegalStateException("empty body")
            var event = "message"
            while (!source.exhausted()) {
                val line = source.readUtf8Line() ?: break
                if (line.isEmpty()) {
                    event = "message"
                    continue
                }
                if (line.startsWith("event: ")) {
                    event = line.removePrefix("event: ")
                    continue
                }
                if (line.startsWith("data: ")) {
                    val payload = line.removePrefix("data: ")
                    parse(event, payload)?.let { frame ->
                        emit(frame)
                        if (frame is AIFrame.Finished || frame is AIFrame.Err) return@use
                    }
                }
            }
        }
    }.flowOn(Dispatchers.IO)

    private fun parse(event: String, payload: String): AIFrame? {
        val obj = runCatching { json.parseToJsonElement(payload) as JsonObject }
            .getOrNull() ?: return null
        val turnID = obj["turn_id"]?.jsonPrimitive?.contentOrNull ?: ""
        return when (event) {
            "token" -> AIFrame.Token(
                turnID = turnID,
                text = obj["text"]?.jsonPrimitive?.contentOrNull ?: "",
            )
            "tool_call" -> AIFrame.ToolCall(
                turnID = turnID,
                callID = obj["call_id"]?.jsonPrimitive?.contentOrNull ?: "",
                name = obj["name"]?.jsonPrimitive?.contentOrNull ?: "",
                argsJSON = obj["args_json"]?.jsonPrimitive?.contentOrNull ?: "",
            )
            "tool_result" -> AIFrame.ToolResult(
                turnID = turnID,
                callID = obj["call_id"]?.jsonPrimitive?.contentOrNull ?: "",
                resultJSON = obj["result_json"]?.jsonPrimitive?.contentOrNull ?: "",
                isError = obj["is_error"]?.jsonPrimitive?.boolean ?: false,
            )
            "finished" -> AIFrame.Finished(
                turnID = turnID,
                stopReason = obj["stop_reason"]?.jsonPrimitive?.contentOrNull ?: "",
                tokensIn = obj["tokens_in"]?.jsonPrimitive?.intOrNull ?: 0,
                tokensOut = obj["tokens_out"]?.jsonPrimitive?.intOrNull ?: 0,
            )
            "error" -> AIFrame.Err(
                turnID = turnID,
                message = obj["error"]?.jsonPrimitive?.contentOrNull ?: "unknown error",
            )
            else -> null
        }
    }
}
